core/src/main/scala/kafka/zk/KafkaZkClient.scala

```diff
@@ -1253,31 +1264,34 @@ class KafkaZkClient(zooKeeperClient: ZooKeeperClient, isSecure: Boolean) extends
   }
 
   private def retryRequestsUntilConnected[Req <: AsyncRequest](requests: Seq[Req]): Seq[Req#Response] = {
-    val remainingRequests = ArrayBuffer(requests: _*)
-    val responses = new ArrayBuffer[Req#Response]
-    while (remainingRequests.nonEmpty) {
-      val batchResponses = zooKeeperClient.handleRequests(remainingRequests)
-
-      // Only execute slow path if we find a response with CONNECTIONLOSS
-      if (batchResponses.exists(_.resultCode == Code.CONNECTIONLOSS)) {
-        val requestResponsePairs = remainingRequests.zip(batchResponses)
-
-        remainingRequests.clear()
-        requestResponsePairs.foreach { case (request, response) =>
-          if (response.resultCode == Code.CONNECTIONLOSS)
-            remainingRequests += request
-          else
-            responses += response
-        }
+    val startNs = time.nanoseconds
+    try {
+      val remainingRequests = ArrayBuffer(requests: _*)
+      val responses = new ArrayBuffer[Req#Response]
+      while (remainingRequests.nonEmpty) {
+        val batchResponses = zooKeeperClient.handleRequests(remainingRequests)
+
+        // Only execute slow path if we find a response with CONNECTIONLOSS
+        if (batchResponses.exists(_.resultCode == Code.CONNECTIONLOSS)) {
+          val requestResponsePairs = remainingRequests.zip(batchResponses)
+
+          remainingRequests.clear()
+          requestResponsePairs.foreach { case (request, response) =>
+            if (response.resultCode == Code.CONNECTIONLOSS)
+              remainingRequests += request
+            else
+              responses += response
+          }
 
-        if (remainingRequests.nonEmpty)
-          zooKeeperClient.waitUntilConnected()
-      } else {
-        remainingRequests.clear()
-        responses ++= batchResponses
+          if (remainingRequests.nonEmpty)
+            zooKeeperClient.waitUntilConnected()
+        } else {
+          remainingRequests.clear()
+          responses ++= batchResponses
+        }
       }
-    }
-    responses
+      responses
+    } finally latencyMetric.update(TimeUnit.NANOSECONDS.toMillis(time.nanoseconds - startNs))
```

Comment:
Would be interested in thoughts on whether this is what we want to measure for the existing metric. cc @rajinisivaram @junrao

Reply:
This now measures the latency for a batch of requests. So, if the batch is large, the latency could be long, but that doesn't necessarily mean a performance issue in ZK. Another way is to do that in ZookeeperClient and track the latency for individual request within a batch. That way, the measured latency may be closer to what we want given that we have a bound on inflight requests.

Reply:
Yes, I thought of that. My concern was that we may do a lot of System.nano calls. For the Kafka protocol, we tend to have request level metrics even though they are batched.

Another possibility is to divide the time by the number of items in the batch. Thoughts?

Reply:
Yes, that's probably a good compromise.

Reply:
OK, I'll go with this approach for the initial version so that we don't lose the metric altogether. I'll file a JIRA to consider the more accurate version after some performance testing.

Reply:
I looked into this some more and this number could be misleading so I think it's better to do it per request as you had originally suggested. The next question is what to measure for a single request. Is the queuing time included? It seems like it might be useful to have 2 metrics, one for the queuing time (a new metric) and one for the time between send and receive (the existing metric). What do you think?

- Time between request arrival and request send (i.e. queue time)
- Time between request send and completion (current metric)